design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Support for pure calculation shaders.  Useful for a variety of things, but think particles.
+
+#### Benefits
+We'll be able to:
+- Create entirely new shaders to do new and different things.
+- Replace existing shaders without having to recompile the whole set.
+- Let developers communicate with their own shaders.
+
+#### For GLES we’ll need to:
+- Push aside the existing Program hierarchy and rename it with a GLES extension.
+- Allow for named uniforms for new Programs<br>We can probably ignore that requirement for the existing shaders and just wrap them
+
+### Shader(Program) Registry
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+Once developers can create their own shaders, we’ll need a way of managing them, thus the Shader Registry.  This registry will associate a given shader (Program) with a name.  Basic functionality will include:
+- Add a new shader by name
+- Replace an existing shader by name
+- Query a shader by name
+- Publish the well known names of the core shaders for the basic functionality
+
+Layers already go through a level of indirection to instantiate or reference their associated programs.  The shader registry could be inserted into that get*LayerPrograms() pattern fairly easily.
+
+![Shader Registry](resources/fig1.png)
+
+The GLES version of the Shader Registry can encapsulate the simple (but fairly weird) logic of shader source lookup that exists now.
+
+#### The way it is now
+
+The layers more or less pull their shaders out of a compressed block of text.  Clever, but difficult to access.
+
+#### Required Changes
+
+We'll need to make a Shader Registry class, of course, but that's fairly simple.  It should be thread safe, but you're mostly going to be using it on the main thread.
+
+The layers should pull their shaders out of the shader on demand.  This will give developers a chance to replace or add shaders before they're needed.
+
+### Rendering Passes
+
+_Addresses core functionality [#5](#core)._
+
+Modern graphics pipelines use more than one pass to create a visual representation.  It's actually pretty unusual to have a single rendering pass in your toolkit.  These are used for a dizzying array of purposes, including lightning effects, shadows, atmosphere and many others.  We (Wet Dog Weather) use separate rendering passes to draw data intermingled with map elements.
+
+Without getting too deep into specifics, early rendering passes allow the developer to use the power of the rasterizer for their own data.  Later rendering passes are typically used to decorate the map with effects.
+
+#### The way it is now:
+The [low level rendering logic](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/renderer_impl.cpp) in the toolkit seems to only support one explicit rendering pass.  
```

Comment:
Looking at the RenderOrchestrator and RenderTree I think they can largely stay intact.  But there's no good reason not to do complete setup on other threads.  I'd have to look more closely at this as I'm not clear how much use the toolkit makes of shared GL contexts.

Some of this might get done in other parts of the rework, but I'm not sure we can make it a hard requirement for this phase.  Curious what people think.

Reply:
From what I've seen doing iOS and server-side (linux) integrations, it makes zero use of shared contexts.

Reply:
That's a big miss, particularly on Android.  We make heavy, heavy use of shared contexts in WhirlyGlobe and it makes for some great threading.

For the scope of this effort, though, we're focusing on making it possible to do Metal (and Vulkan, etc).  After that's done it might be good to revisit OpenGL for Android and see what's possible.